New implementation of license compatibility

hesa opened the pull request:
These commits provides a new implementation of license compatibility checks for projects (with dependencies). It (hopefully) includes updated doc and tests as well.

The goals of this rewrite have been to:
* increase (ehum) readability and maintainability
* increase speed

For more info, see: README.md and SETTINGS.md

priv-kweihmann committed:
test/comp_matrix: fix test with LGPL and GPL

as LGPL2 is only compatible one way with GPL2

Signed-off-by: Konrad Weihmann <kweihmann@outlook.com>

priv-kweihmann committed:
requirements: bump osadl-matrix

to latest available

Signed-off-by: Konrad Weihmann <kweihmann@outlook.com>

priv-kweihmann committed:
requirements: bump osadl-matrix

to latest available

Signed-off-by: Konrad Weihmann <kweihmann@outlook.com>

hesa committed:
Add format functions for new functionality

remove old functions

hesa committed:
Format: Add format functions for new functionality

Remove old functions

hesa committed:
Project: Add file to read project (flict, SPDX)

hesa committed:
Project: Add file to read project (flict, SPDX)

hesa committed:
Config: Remove old settings, add for new features

hesa committed:
Config: Remove Unused settings, add new for new features

hesa committed:
Config: Remove old settings, add for new features

hesa committed:
Config: Remove old settings, add new for features

hesa committed:
Log: Set default log level to DEBUG

hesa committed:
Test: Add file with commands to check cli

hesa committed:
Test: Add file with commands to check cli

hesa committed:
fix typo

hesa committed:
Config: Fix typo

hesa committed:
Config: Fix typo

hesa committed:
Main: Switch from old to new implenentation

hesa committed:
Impl: Switch from old to new implementation

hesa committed:
New implementation of flict features

hesa committed:
flictlib: New implementation of flict features

hesa committed:
License: Add new license module

hesa committed:
License: Add new license module

hesa committed:
Main: Replace command line options for new functionality

hesa committed:
Compatibility: Add new compat module

hesa committed:
new compat module

hesa committed:
Test: Add test file with additions to matrix

hesa committed:
Test: Add test file with additions to matrix

hesa committed:
Devel: Add test file with additions to matrix

hesa committed:
Alias: add new alias file (JSON)

hesa committed:
Alias: Add new alias file

hesa committed:
sample SPDX 2.2 files

hesa committed:
Example: Add sample SPDX 2.2 files

hesa committed:
Example: Add sample SPDX 2.2 files

hesa committed:
Test: Fix paths

hesa committed:
Test: Fix path typo

hesa committed:
Test: Fix path typo

hesa committed:
Test: Fix path typo

hesa committed:
path typo

hesa committed:
Unused files: Remove unused files

hesa committed:
Reuse: Remove old files

hesa committed:
Documentation: Add updates for new options and features

hesa committed:
Doc: Add updates for new options and features

hesa committed:
Doc: Add updates for new options and features

hesa committed:
updates for new options and features

hesa committed:
Doc: Add updates for new options and features

hesa committed:
Doc: Add updates for new options and features

hesa committed:
updates for new options and features

hesa committed:
Documentation: Add updates for new options and features

hesa committed:
Test: Fix indent

hesa committed:
Config: Fix indent

hesa committed:
Test: Fix indent

hesa committed:
Main: Fix indent

hesa committed:
Config: Fix space

hesa committed:
Main: remove old options

hesa committed:
remove old options

hesa committed:
Main: Remove old options

hesa committed:
Test: Update tests for new functionality

hesa committed:
Test: Update tests for new functionality

hesa committed:
Test: Update tests for new functionality

hesa committed:
Test: Update tests for new functionality

hesa committed:
Unused files: Remove unused files

hesa committed:
Unused files: Remove unused files

hesa committed:
Config: Fix alias file typos

Requirements: fix osadl_matrix version

hesa committed:
fix osadl_matrix version

priv-kweihmann commented on SETTINGS.md:
````diff
@@ -6,126 +6,151 @@ SPDX-License-Identifier: GPL-3.0-or-later
 
-```spdx``` - the license (SPDX short name) this later definition is valid for
+```
+$ flict -of text verify -il MIT AND BSD-3-Clause -ol X11
````
I think we should propose to quote the license string, so it would be `"MIT AND BSD-3-Clause"` for something like this

priv-kweihmann commented on SETTINGS.md:
````diff
@@ -6,126 +6,151 @@ SPDX-License-Identifier: GPL-3.0-or-later
+```
+$ flict -of text verify -il MIT AND BSD-3-Clause -ol X11
+Yes
+$ flict -of text -ldf example-data/denied-licenses.json verify  -il MIT AND BSD-3-Clause -ol X11
````
Same here - pls quote the license string

priv-kweihmann commented on SETTINGS.md:
```diff
@@ -6,126 +6,151 @@ SPDX-License-Identifier: GPL-3.0-or-later
+most preferred license. By default flict counts how many other
+licenses each license is compatible with. The more licenses a license
+is compatible with the more preferred it will be. If two licenses have
+the same number of compaitbilities alpabetical order will be used to
```
typo `compaitbilities`

priv-kweihmann commented on flict/__main__.py:
```diff
@@ -151,18 +149,23 @@ def parse():
     parser_v.set_defaults(which="verify", func=verify)
+    parser_v.add_argument('--outbound-license', '-ol', type=str, dest='out_license', help='Outbound license', default=None)
+    parser_v.add_argument('--inbound-license', '-il', type=str, nargs='+', dest='in_license_expr', help='Inbound license(s)', default=[])
+    parser_v.add_argument('--sbom', '-s', type=str, dest='verify_sbom', help='')
```
I'd prefer if you either remove the help options or fill it with something meaningful

priv-kweihmann commented on flict/flictlib/alias.py:
```diff
@@ -0,0 +1,62 @@
+def _read_aliases(alias_file=FLICT_ALIAS_FILE):
+    global alias_object
+    with open(alias_file) as fp:
+        alias_object = json.load(fp)
```
Some error handling would be appreciated - maybe using the error class

priv-kweihmann commented on flict/flictlib/format/dot_format.py:
```diff
@@ -128,3 +111,5 @@ def _compat_to_dot(left, comp_left, right, comp_right):
             {_print_compare_line(right, left, '[color="gray", style="dotted"]')}
             """
+    else:
+        print("WHAT???")
```
???? - this should be using logging - looks like a assertion to me

priv-kweihmann commented on flict/flictlib/format/markdown_format.py:
```diff
@@ -10,71 +10,183 @@
+            if description != "":
+                output.append(f"*Description*: {description}\n")
+
+            if identified_licenses == []:
```
`if not any(identified_licenses)` - it's faster

hesa replied:
if not description:
   :)

priv-kweihmann commented on flict/flictlib/format/markdown_format.py:
```diff
@@ -10,71 +10,183 @@
+
+            output.append(self.package_name_header(package['name']))
+            description = package.get('description', "")
+            if description != "":
```
`if not description`

priv-kweihmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,216 @@
+
+        self.compatibility = CompatibilityFactory.get_compatibility(license_db)
+
+        if licenses_preferences is None or licenses_preferences == []:
```
`if not licenses_preferences`

priv-kweihmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,216 @@
+            return expr
+        else:
+            # TODO: raise exception
+            logging.debug("whats???? ")
```
yes logging.exception + raise a FlictError

priv-kweihmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,216 @@
+        if updated:
+            return CompatibilityStatus.LICENSE_COMPATIBILITY_COMPATIBLE.value
+        return CompatibilityStatus.LICENSE_COMPATIBILITY_INCOMPATIBLE.value
+        # TODO: raise exception
```
after a the return? I don't think so :-)

hesa replied:
:(

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,228 @@
+    def _flatten_packages(self, packages):
+        package_list = []
+
+        for pkg in packages.items():
```
a better pattern for a dict is `for k, v in packages.items()` then you can use `k` (for the key) and `v` for the value.
That looks much more readable than the tuple with an index handling below

hesa replied:
:)

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,228 @@
+            raise FlictError(ReturnCodes.RET_INVALID_PROJECT, f'File "{project_file}" does not contain valid JSON data')
+        except (FileNotFoundError, IsADirectoryError):
+            raise FlictError(ReturnCodes.RET_FILE_NOT_FOUND, f'File "{project_file}" could not be found or is a directory')
+        except:
```
here it'd be nice to show the actual exception in the message...
```
except Exception as e:
   raise FlictError(ReturnCodes.RET_INVALID_PROJECT, f'File "{project_file}" could not be parsed: {e}.')
```

priv-kweihmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,228 @@
+
+        if spdx_version.startswith("2.2"):
+            return self._read_spdx_2_2(only_packages=None)
+        # TODO: raise exception with meaningful hint that SPDX version not supported
```
pls fix that here - maybe with a generic FlictError, but returning with nothing where one expects a return is nothing we should merge like this

priv-kweihmann commented on flict/flictlib/utils.py:
```diff
@@ -0,0 +1,28 @@
+        'osVersion': uname.version,
+        'machine': uname.machine,
+        'host': uname.nodename,
+        'user': getpass.getuser(),
```
why not `os.environ.get('USER')` ?

priv-kweihmann commented on flict/impl.py:
```diff
@@ -3,137 +3,108 @@
-        except:
-            raise FlictError(ReturnCodes.RET_INVALID_EXPRESSSION,
-                             f'Could not parse license expression: {self._args.license_expression}')
+        if self._args.out_license and self._args.in_license_expr != []:
```
` and any(self._args.in_license_expr)` or `and self._args.in_license_expr`

hesa replied:
You mean, replace:
`        if self._args.out_license and self._args.in_license_expr != []:`

with:
`        if self._args.out_license and any(self._args.in_license_expr) or and self._args.in_license_expr`

?

priv-kweihmann replied:
yeah you got me confused with this `!= []` pattern - just replace that and it should be good

hesa replied:
Oh dear ... I missed that one of your "or" was not code :)

priv-kweihmann commented on tests/test_compat.py:
```diff
@@ -1,24 +1,87 @@
 
-import json
-import pytest
+import unittest
```
why the fallback on legacy `unitest` - I'd prefer sticking with way mightier `pytest`

hesa replied:
Suggest switching to that for all files in separate branch.

priv-kweihmann commented on tests/test_compat.py:
```diff
@@ -1,24 +1,87 @@
-    assert '{"compatibilities": [{"license": "MIT", "licenses": []}]}' == ret
+    def test_multiple_license(self):
+        license_list = self.choser.list()
+        print("wello: " + str(license_list))
```
print statement won't make it into the testing log, which renders them a bit useless.
pls use `logging.warning` - so in case the test fails it will be automatically shown in the console output of pytest

priv-kweihmann commented on tests/test_simplified.py:
```diff
@@ -10,7 +11,9 @@
     args = ArgsMock(license_expression=expression)
     ret = FlictImpl(args).simplify()
-    assert result == ret
+    print("ret:    " + str(ret))
```
Same here with the print statements

priv-kweihmann commented on tests/test_verify.py:
```diff
@@ -1,31 +1,61 @@
 
-import os
-import pytest
+import unittest
```
Same here - why unittest instead of pytest?

priv-kweihmann commented on tests/test_verify.py:
```diff
@@ -1,31 +1,61 @@
+        compat = package['compatibility'][0]
+
+        compats = package['compatibility']
+        self.assertTrue(len(compats)==1)
```
`self.assertEqual(len(compats), 1)`

priv-kweihmann commented on tests/test_verify.py:
```diff
@@ -1,31 +1,61 @@
+        # freetype specific check
+        package = verification['packages'][0]
+        compats = package['compatibility']
+        self.assertTrue(len(compats)>1)
```
`self.assertEqual(len(compats), 1)`

hesa committed:
Test: Add (c) and license info

hesa committed:
add (c) and license info

hesa committed:
SETTINGS: quote license string

hesa committed:
Settings: Fix typo

hesa committed:
Main: Improve help text for verify command

hesa committed:
Alias: Make class out of functions

hesa committed:
Main: fix type for alias_file

hesa committed:
Arbiter: pass alias file to LicenseCompatibility

hesa committed:
Compatibility: Switch to using Alias class

hesa committed:
Compatibility: Switch to using Alias class

hesa committed:
Compatibility: Switch to using Alias class

hesa committed:
Return codes: add invalid alias file

hesa committed:
Arbiter: pass alias file to Arbiter

hesa committed:
Test: Add alias_file

hesa committed:
Test: Clean up print statements

hesa committed:
Test: Use proper assert

hesa committed:
Format: Replace printout with exception

hesa committed:
Format: Replace str comparison with not

hesa committed:
Lic-Com: Raise exceptions on internal errors

Fix misc comparisons

hesa committed:
Project: Raise error if SPDX version unsupported

Fix loop over packages

hesa committed:
Utils: Replace getpass with os calls

hesa committed:
Project: Add cause when raising exception

hesa committed:
Fix indent

hesa committed:
Fix indent

hesa committed:
resolve conflict

hesa committed:
Merge branch 'hesa-hesa-use-license_compatibility'

hesa committed:
Setup: Remove blank lines

hesa committed:
Impl: make more pythonic

hesa committed:
Impl: make more pythonic

hesa committed:
Example: Add example files

hesa committed:
Example: Add example files

hesa committed:
Merge branch 'vinland-technology:main' into main

hesa committed:
Merge branch 'hesa-use-license_compatibility' of github.com:hesa/flict into hesa-use-license_compatibility

hesa committed:
Test: remove redirect to dev null

hesa committed:
Flict: new license compatibility implementation

For more info, see: README.md and SETTINGS.md

jens-erdmann commented on flict/__main__.py:
```diff
@@ -175,15 +178,19 @@ def parse():
+#    parser_li.add_argument('-r', '--relicensing',
+#                           dest='list_relicensing',
+#                           action='store_true',
+#                           help='List relicensing information')
```
If the options above are not needed, they should be removed. We should not introduce more dead code.

jens-erdmann commented on flict/flictlib/alias.py:
```diff
@@ -0,0 +1,60 @@
+    def format_aliased(self, lic):
+        alias = self.replace_aliases(lic)
+        if alias != lic:
+            return lic + " (aliased to " + alias + ")"
```
`return f'{lic} (aliased to {alias})'`

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+            compat = self.inbounds_outbound_check(outbound_license, [inbound_license])
+            compats.append(compat)
+
+        return compats
```
`return [self.inbounds_outbound_check(outbound_license, [inbound_license]) for outbound_license in licenses]`

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+        if not compatible_license:
+            logging.debug("None such ... probably because complex")
+            # TODO: investigate if this can happen at all
+            sys.exit(1)
```
This should be an exception instead of `sys.exit(1)`.

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+            compatible = compatible_license[0]
+
+        for dep_info in dep_infos:
+            # Get the corresponding compatiblity for outbound_license for dep_info
```
This should be a separate function.

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+
+        logging.debug("ALL LICENSES OUTBOUND")
+        #
+        # Let's loop through all the other licenses in the combined work
```
Why?

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
```
Shebangs are not required here. It is enough to have it in the main file.

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+            'compatibility': compats
+        }
+
+    def verify(self, project, licenses=None):
```
This does too many things at once IMHO.

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+                 expr: license expr with inbound license (e.g. "MPL-2.0 OR MIT")
+        """
+        compats = self.license_compatibility.inbounds_outbound_compatibility(outbound, expr)
+        return compats
```
`return self.license_compatibility.inbounds_outbound_compatibility(outbound, expr)`

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,226 @@
+                 ibound: inbound license (e.g. "MPL-2.0")
+        """
+        compats = self.license_compatibility.inbound_outbound_compatibility(outbound, inbound)
+        return compats
```
`return self.license_compatibility.inbound_outbound_compatibility(outbound, inbound)`

jens-erdmann commented on flict/flictlib/format/format.py:
```diff
@@ -49,3 +43,22 @@ def format_translation_information(self, license_handler):
+            #print("--compare: " + str(outbound_lic) + " <--> " + outbound)
+
+            if outbound_lic == outbound:
+                #print("--found: " + str(outbound_lic) + " <--> " + outbound)
```
Dead code should be removed.

jens-erdmann commented on flict/flictlib/format/markdown_format.py:
```diff
@@ -10,71 +10,183 @@
+            main_license = compat['license']
+            result.append(self._output_compat_markdown_licenses(main_license, compat))
+
+        return result
```
Why is this returning an array of strings? This could be refactored into a static multiline string.

jens-erdmann commented on flict/flictlib/format/markdown_format.py:
```diff
@@ -10,71 +10,183 @@
+            if not dep_identified_license:
+                output.append(self.problem_header())
+                output.append("Could not conclude an outbound license for the combined work.\n")
+        return "\n".join(output)
```
Same here. The structure of the string could be pre-defined and the actual variables should be provided by separate functions.

jens-erdmann commented on flict/flictlib/format/text_format.py:
```diff
@@ -79,3 +77,64 @@ def format_policy_report(self, policy_report):
+                return lic['compatible_left'].replace("true", "Yes").replace("false", "No")
+
+    def _format_lic(self, lic):
+        str_size = "{0: <" + str(self.col_size) + "}"
```
Good ol C code again.
`f'{0: <{self.col_size}}'`

jens-erdmann commented on flict/flictlib/format/text_format.py:
```diff
@@ -79,3 +77,64 @@ def format_policy_report(self, policy_report):
+    def _format_line(self):
+        return self._format_lic("-" * (self.col_size - 1))
+
+    def format_compats(self, compats):
```
This should be broken down as well. Too many things at the same time.

jens-erdmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,217 @@
+
+        compats = self._inbounds_outbound_check(parsed_outbound, parsed)
+
+        return compats
```
Agani: compats should be returned directly.

jens-erdmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,217 @@
+
+    def _inbounds_outbound_check(self, outbound, expr):
+        logging.debug(f"_inbounds_outbound_check({outbound}, {expr})")
+        if self.license_parser.is_license(expr):
```
Both `if` and `elif` path should be a function.

jens-erdmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,217 @@
+
+            expr['outbound'] = outbound
+            expr["allowed"] = allowed_summary
+            expr['check'] = 'inbounds_outbound'
```
Lookup of keys is expensive. `expr` should be created inline instead of being expanded by key value after key value.

jens-erdmann replied:
Redacting my statement. I thought, this dict would be newly created in here.

jens-erdmann commented on flict/flictlib/license.py:
```diff
@@ -1,382 +1,304 @@
+LICENSE_SYMBOL = "LicenseSymbol"
+LICENSE_WITH_SYMBOL = "LicenseWithExceptionSymbol"
+LICENSE_EXPRESSION_OR = "OR"
+LICENSE_EXPRESSION_AND = "AND"
```
These should be an enum to make it easier to identify.

jens-erdmann commented on flict/flictlib/license.py:
```diff
@@ -1,382 +1,304 @@
+                #logging.debug("2-: " + expr[1:size+index])
+                return size + index
+
+        # TODO: raise exception
```
Please do.

jens-erdmann commented on flict/flictlib/license.py:
```diff
@@ -1,382 +1,304 @@
+                #if ")" in expr[size+1:]:
+                #    index = expr[size+1:].index(")")
+
+                #logging.debug("2-: " + expr[1:size+index])
```
Dead code should be removed.

jens-erdmann commented on flict/flictlib/format/markdown_format.py:
```diff
@@ -10,71 +10,183 @@
+            output.append(self.package_name_header(package['name']))
+            description = package.get('description', "")
+            if not description:
+                output.append(f"*Description*: {description}\n")
```
Isn't this the wrong way around? If `description` is empty, why add it to the list?

jens-erdmann commented on flict/flictlib/project/__init__.py:
```diff
@@ -0,0 +1,4 @@
+
+# SPDX-FileCopyrightText: 2022 Henrik Sandklef
```
Copyright on an empty file? Bold move

jens-erdmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,229 @@
+
+            dep_list = []
+            for dep in tmp_dict.values():
+                dep_list.append(dep)
```
This is doing things almost trice. Shouldn't do `dep_list = tmp_dict.values()` do basically the same?

jens-erdmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,229 @@
+                    if dep_package_name == _pkg_name:
+                        packages[top_package]['dependencies'].append({_pkg_name: _pkg[1]})
+                    else:
+                        pass
```
This is not needed.

jens-erdmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,229 @@
+                    else:
+                        pass
+
+        ret = {
```
This should be returned directly.

hesa committed:
Main: Remove dead code

hesa committed:
Alias: Use f strings instead of +

hesa committed:
Arbiter: Replace exit call with exception

Use list comprehensions
Move loop to separate function

hesa committed:
Format: Fix conditional description inclusion

Remove shebang

hesa committed:
Project: Remove (c) statement from empty file

hesa committed:
Format: Replace tedious loop construct

Remove useless code (pass)

hesa committed:
Project: Return value directly (no temp var)

jens-erdmann commented on flict/impl.py:
```diff
@@ -3,137 +3,109 @@
-        project = Project(None, self._license_handler, licenses)
-        return Report(project, self._compatibility)
+    def merge_license_db(self):
+        arbiter = self._get_arbiter()
```
If every method needs access to an arbiter, an arbiter should be part of `__init__`

hesa committed:
remove shebang

hesa committed:
make function inheritable

hesa committed:
remove uselss tests, import from license_parser

hesa committed:
Create license objects in constructors

hesa committed:
Use new License class instead

hesa committed:
Move all parse to license_parser

hesa committed:
Fix typo

hesa committed:
Update README

Add linkt to Konrad's osadl_matrix and matrix.json

hesa committed:
Create arbiter in __init__

hesa committed:
New file to parse licenses

hesa committed:
All: Resolving Jens' comments

hesa committed:
Format: pass license_license

jens-erdmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,217 @@
+                allowed = compat['allowed']
+                allowed_summary = self._update_allowed(op, allowed_summary, allowed)
+                operand['allowed'] = allowed
+                operand['allowed'] = allowed
```
This looks like a doublet to me.

jens-erdmann commented on flict/flictlib/format/text_format.py:
```diff
@@ -79,3 +77,64 @@ def format_policy_report(self, policy_report):
+        inner.append(self._format_line())
+#        for lic in license_list:
+#            inner.append(self._format_line())
+        ret.append("".join(inner))
```
` ret.append("".join(self._format_line()))`

jens-erdmann commented on flict/flictlib/project/reader.py:
```diff
@@ -0,0 +1,229 @@
+        try:
+            with open(project_file, 'r') as f:
+                project_data = json.load(f)
+                project = project_data['project']
+
+                return {
+                    "project_name": project['name'],
+                    "packages": [
+                        {
+                            'name': project['name'],
+                            'version': project['version'],
+                            'license': project['license'],
+                            'description': '',
+                            'dependencies': project['dependencies']
+                        }
+                    ]
+                }
```
This part could be moved out of the try catch block.

hesa replied:
will keep this to create a flict exception if the project file is faulty (e.g. missing key)

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,220 @@
+                index = package_info['outbound_licenses_aliased'].index(chosen_alias)
+                chosen_license = outbound_licenses[index]
+                package_info['outbound_license'] = chosen_license
+                package_info['outbound_license_aliased'] = chosen_alias
```
This part is still confusing. In the previous lines you get outbound licenses and outbound licenses aliased. None of them matter because you override them either way in this section. Why write temporary data into `package_info`?

hesa replied:
I have added some comments on `outbound_license` vs `outbound_licenses` (singular vs plural) hoping that they make it easier to follow

hesa replied:
Example:
```
      "outbound_licenses": [
        "GPL-3.0-only", "AGPL-3.0-only"
      ],
      "outbound_licenses_aliased": [
        "GPL-3.0-only", "AGPL-3.0-only"
      ],
      "outbound_license": "GPL-3.0-only",
      "outbound_license_aliased": "GPL-3.0-only"
```

The outbound licenses (plural) are listed:   "GPL-3.0-only", "AGPL-3.0-only"
..as well as the chosen (singular) license: , "GPL-3.0-only"

hesa replied:
The idea is to give a compliance person a chance to see which license was chosen as outbound and from what licenses that was chosen from,

jens-erdmann replied:
Ah my sloppy reading strikes again. Maybe that naming could be improved in the future.

jens-erdmann commented on flict/flictlib/format/markdown_format.py:
```diff
@@ -87,27 +186,14 @@ def _compat_to_markdown(left, comp_left, right, comp_right):
+        comp_right = lic['compatible_right']
+        compat_text = _compat_to_markdown(
+            main_license, comp_left, inner_license, comp_right)
+        result.append(main_license + " " + compat_text + " " + inner_license + "\n\n")
```
F-String

jens-erdmann commented on flict/flictlib/format/text_format.py:
```diff
@@ -79,3 +77,67 @@ def format_policy_report(self, policy_report):
+        #print(str(compat))
+        for lic in compat['licenses']:
+            if lic['license'] == license_name:
+                return lic['compatible_left'].replace("true", "Yes").replace("false", "No")
```
These two have a big overlap and are not related to the format. How come we do not have this in other formats as well?

jens-erdmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,215 @@
+        #expr['compatibility_info'] = compat
+        expr[COMPATIBILITY_TAG] = compat_summary
+
+        expr['outbound'] = outbound
```
This gets set twice.

hesa replied:
You know, one for the road ;)

jens-erdmann commented on flict/flictlib/lic_comp.py:
```diff
@@ -0,0 +1,215 @@
+            allowed = compat['allowed']
+            allowed_summary = self._update_allowed(op, allowed_summary, allowed)
+            operand['allowed'] = allowed
+            operand['allowed'] = allowed
```
This gets set twice.

hesa committed:
Example: Add version info

hesa committed:
Arbiter: Add doc

hesa committed:
Format: Move common code to top class

hesa committed:
Format: Use f-strings

hesa committed:
Misc: chose -> choose

hesa committed:
Project: Detect JSON error and create exception

hesa committed:
Misc: chose -> choose

jens-erdmann commented on flict/flictlib/arbiter.py:
```diff
@@ -0,0 +1,224 @@
+                package_info['outbound_license'] = chosen_license
+                package_info['outbound_license_aliased'] = chosen_alias
+
+            package_infos.append(package_info)
```
```
            all_licenses.update(licenses)

            package_info = self._package_info(package, licenses)

            dep_infos = [self._package_info(dep, licenses) for dep in package.get('dependencies', [])]

            # Get a list of the outbound licenses for all packages
            outbound_licenses = self._top_package_license(package['license'], licenses, package_info, dep_infos)

            # Get the alias for all outbound licenses
            outbound_licenses_aliased = [self.license_compatibility.replace_aliases(lic) for lic in outbound_licenses]

            # Identify single outbound (chosen) license (from the aliased outbound licenses)
            chosen_alias = self.license_compatibility.chose_license(outbound_licenses_aliased)
            if chosen_alias is None:
                chosen_license = None
            else:
                # Find the index of the aliased license
                # and use that to identify the original (not aliased) license
                index = outbound_licenses_aliased.index(chosen_alias)
                chosen_license = outbound_licenses[index]

            package_infos.append({
                'dependencies': dep_infos,
                'outbound_licenses': outbound_licenses,
                'outbound_licenses_aliased': outbound_licenses_aliased,
                'outbound_license': chosen_license,
                'outbound_license_aliased': chosen_alias
            })
```

hesa replied:
```
            package_info.update({
                'dependencies': dep_infos,
                'outbound_licenses': outbound_licenses,
                'outbound_licenses_aliased': outbound_licenses_aliased,
                'outbound_license': chosen_license,
                'outbound_license_aliased': chosen_alias
            })
            package_infos.append(package_info)
```

hesa committed:
Use list comprehension, skip temp var

hesa committed:
Arbiter: Add package_info

hesa committed:
A new implementation of license compatibility checks

* for projects (with dependencies)
* with a new report format

hesa committed:
New compatibility implementation

Alias:

* Add new alias file (JSON)

* Updates for new options and features

* Remove old options

* Use f strings instead of +

Arbiter:

* Pass alias file to LicenseCompatibility

* Replace exit call with exception

* Use list comprehensions

* Move loop to separate function

Compatibility:

* Add new compat module

Config:

* Remove old settings, add for new features

Documentation:

* Quote license string in SETTINGS

* Updates for new options and features

* Add link to Konrad's osadl_matrix and matrix.json (README)

* Add sample SPDX 2.2 files

Format:

* Add format functions for new functionality

* Fix conditional description inclusion

* Remove unused functions

* Replace printout with exception

* Replace str comparison with not

* Use f-strings

Impl:

* Switch from old to new implementation

* Make more pythonic

License:

* Add new license module

* Move all parse to license_parser

License Compatibiility:

* Raise exceptions on internal errors

* For more info, see: README.md and SETTINGS.md

* Create license objects in constructors

* Use new License class instead

Log:

* Set default log level to DEBUG

Main:

* Replace command line options for new functionality

* Improve help text for verify command

* Make class out of functions

* Remove dead code

Project:

* Add file to read project (flict, SPDX)

* Raise error if SPDX version unsupported

* Add cause when raising exception

* Remove (c) statement from empty init file

* Replace tedious loop construct with comprehensions

* Return value directly (no temp var)

Requirements:

* Correct osadl_matrix version

Return codes:

* Add invalid alias file

Reuse:

* Remove old files

Setup:

* Remove blank lines

Test:

* Add file with commands to check cli

* Update tests for new functionality

* Add (c) and license info

* Add test file with additions to matrix

* Unused files: Remove unused files

* Clean up print statements

* Use proper asserts

* Remove redirect to dev null

Utils:

* Replace getpass with os calls

hesa commented:
Alias:

* Add new alias file (JSON)

* Updates for new options and features

* Remove old options

* Use f strings instead of +

Arbiter:

* Pass alias file to LicenseCompatibility

* Replace exit call with exception

* Use list comprehensions

* Move loop to separate function

Compatibility:

* Add new compat module

Config:

* Remove old settings, add for new features

Documentation:

* Quote license string in SETTINGS

* Updates for new options and features

* Add link to Konrad's osadl_matrix and matrix.json (README)

Examples:

* Add sample SPDX 2.2 files

Format:

* Add format functions for new functionality

* Fix conditional description inclusion

* Remove unused functions

* Replace printout with exception

* Replace str comparison with not

* Use f-strings

Impl:

* Switch from old to new implementation

* Make more pythonic

License:

* Add new license module

* Move all parse to license_parser

License Compatibiility:

* Raise exceptions on internal errors

* For more info, see: README.md and SETTINGS.md

* Create license objects in constructors

* Use new License class instead

Log:

* Set default log level to DEBUG

Main:

* Replace command line options for new functionality

* Improve help text for verify command

* Make class out of functions

* Remove dead code

Project:

* Add file to read project (flict, SPDX)

* Raise error if SPDX version unsupported

* Add cause when raising exception

* Remove (c) statement from empty init file

* Replace tedious loop construct with comprehensions

* Return value directly (no temp var)

Requirements:

* Correct osadl_matrix version

Return codes:

* Add invalid alias file

Reuse:

* Remove old files

Setup:

* Remove blank lines

Test:

* Add file with commands to check cli

* Update tests for new functionality

* Add (c) and license info

* Add test file with additions to matrix

* Unused files: Remove unused files

* Clean up print statements

* Use proper asserts

* Remove redirect to dev null

Utils:

* Replace getpass with os calls
